### ISSUE TITLE
Log lines for an active step are inaccessible

### ISSUE THREAD

**Issue by bogosj** [actions/runner#886]
**Describe the bug**
When opening the "run" page for an active job you can see:

- Gray checkmarks and expandable sections for completed steps and all of their log lines.
- Yellow loading animation on any active steps.
- Circles for the future steps.

The log lines for the active steps will then start streaming in, **but** the prior lines for that step are not there, and there appears to be no way to see them until the step completes.

**To Reproduce**
1. Run an action
2. Visit the run page after a few seconds and try to find the log lines for the currently running step that occurred before you loaded the page.

**Expected behavior**
All of the log lines that were emit before I loaded the page should be visible, then new ones stream in afterwards.

**Comment by bogosj** [comment]
It looks like this is a KI, but I can't find a canonical bug tracking it:

https://github.com/mxschmitt/action-tmate/issues/1#issuecomment-526750603

**Comment by konradpabjan** [comment]
👋 Actions engineer here

So I wouldn't classify this as a bug, but it's just a missing feature.

When a step is running, you can do "inspect element", and see that there is a active websocket that is streaming in new output from the runner. When a step is completed though, a GET request is made to our service that downloads all the logs for step in one go. If a step is in-progress and a user refreshes a page or first loads a page while it is in-progress, the logs that were already streamed to the websocket are effectively gone until the step is completed and we can fetch everything using an internal GET api. 

The solution for us is to somehow cache existing logs that have already been streamed and if a user refreshes a step that is in-progress, fetch the cached logs and display that + append new logs that are coming in through the websocket. 

We'll eventually add this feature (this would be really really nice to have), but at the moment we just haven't been able to pick this up.

**Comment by vadi2** [comment]
This would be great to have because some steps can take a while, or in fact due to error get stuck for a good half hour - and you're sitting there wondering if you should cancel it or not. It comes up often as a problem during development of your workflow!

**Comment by TingluoHuang** [comment]
I believe this is on our backlog.

**Comment by graeme-verticalscope** [comment]
yeah this would be nice to have 😃

**Comment by DavyLandman** [comment]
As an intermediate solution it might be nice to just stream the output that occurs when the user has opened the page? I can understand caching previous output can be complex, but if I just get an output like:

```
WARNING: Truncated output, will be shown after step has finished..
....
new output that the process is generation
```

That could give me an idea if a long running step is stuck (or where it's at by now).

**Comment by ethomson** [comment]
Hi!  This seems like a possibly useful addition.  However, suggestions for GitHub Actions functionality belong in the [GitHub feedback forum](https://github.com/github/feedback/discussions/categories/actions-and-packages-feedback), where the product managers and engineers will be able to review them.  This repository is really only focused on the runner application, which is a very small component of GitHub Actions.

We do have this opened on our backlog, so I'm going to close this issue here.  Thanks again!

**Comment by FranDepascuali** [comment]
This is really needed.
This is the only thing I'm seeing, it has been a long time and I have no way of diagnosing what's happening 
![image](https://user-images.githubusercontent.com/12101394/186014288-600e940b-bc1c-4bcc-af41-e9347aec955d.png)

**Comment by schipperkp** [comment]
I agree with the thought process that this is not nice to have but a must-have. If you're working on debugging a prod deploy that has gone south, you may not be able to cancel the build, but rather you must let it complete the rollback steps.  This means you lose valuable time and information by needing to wait...

**Comment by schipperkp** [comment]
@ethomson  can we get a link to the backlog Item in this thread?

**Comment by umm0n** [comment]
As others have commented, I don't see this as a nice-to-have but rather, an elemental feature. A link to the backlog feature would be nice!

**Comment by julienbonastre** [comment]
Not a fan of bumping for the sake of it, but yes, +1, this is not a "nice to have" but a fundamental feature that should have been part of initial design requirements for a CI solution - if not realtime, at least near-realtime access to logging of job outputs is crucial

**Comment by bitdeft** [comment]
I'm also going to heavily +1 this. It was mentioned that a discussion would need to be opened to express desires for this feature/bugfix to be resolved sooner than later if it's on the backlog. 

I didn't see any when I searched, so I've created this discussion here: 
https://github.com/community/community/discussions/35363

**Comment by 0xF4D3C0D3** [comment]
?? why did this get closed? I think this is not a "nice to have" feature, but rather a "missing" feature. it's been almost three years and I can't still see the log in real time..

**Comment by 0xF4D3C0D3** [comment]
this is not released yet, they seem making a progress for this issue behind the scene. 

<img width="1667" alt="image" src="https://user-images.githubusercontent.com/19762154/206668529-c9b02eca-9f09-48f6-b199-59f42d259481.png">

**Comment by paulogodinhoaq** [comment]
@0xF4D3C0D3 do you believe these are related to this issue? They seem to be about getting the actual Worker logs into our hands, rather than the step its executing. I may be misreading.

**Comment by charlesritchea** [comment]
> @0xF4D3C0D3 do you believe these are related to this issue? They seem to be about getting the actual Worker logs into our hands, rather than the step its executing. I may be misreading. 

I've been following this for awhile I've always understood it to be the step it's executing. Currently the logs often don't show up until after the step finishes. Can be a long time and you have no idea if your step is hung or not

**Comment by paulogodinhoaq** [comment]
> 

Same, this is currently defeating my transition from Jenkins to Actions. 

By my comment I meant that these commits recently made by @AvaStancu may not be related to this issue here, but aiming to help people wanting to watch/debug the runner logs itself. 

Unfortunately, the Issues related to those commits are in a private repo and we can't get context.

**Comment by etaormm** [comment]
I don't think this is being worked on right now...
I have opened the same issue a couple months ago - but no one is assigned to it yet.

https://github.com/actions/runner/issues/2131

**Comment by mcarans** [comment]
This is definitely what I would call a bug and it's rather annoying to not be able to get an idea of where a step has reached particularly when debugging very long-running steps.

**Comment by bryanmacfarlane** [comment]
Hey, want to follow up here from GitHub Actions engineering.

First and foremost just wanted to let everyone know that we hear you.  We're very aware that this is a big ask and one of the big pain points with actions.

Over recent months there's big architectural changes happening as part of larger efforts to set us up for many larger goals; this being one of them.  The services and the key pinnings behind those changes are out in production and being phased in.   We hope to realize this as part of those changes soon. 

Even though the issue is closed (since it's really not primarily a runner issue) we wanted to follow up since this is high activity.  Thanks for the (very clear) feedback here. :)

**Comment by axel-h** [comment]
About three months later and I'm still waiting for a practical solution. It's really a pain when you have PRs triggered test jobs that runs for over an hour producing logs from time to time, but there is no way to have a look at what has been printed so far - unless you capture all log from the beginning.

**Comment by julienbonastre** [comment]
> About three months later and I'm still waiting for a practical solution. It's really a pain when you have PRs triggered test jobs that runs for over an hour producing logs from time to time, but there is no way to have a look at what has been printed so far - unless you capture all log from the beginning.

Anyone have an idea of what the **ACTIVE and OPEN** issue is for tracking the acclaimed changes that would remediate this? So far we are all here watching and waiting on closed issues (this one and https://github.com/actions/runner/issues/891 ) - doesn't bode well for me trying to have optimism on issues that have no priority or have been deemed completed 😞

**Comment by erikahansen** [comment]
Maybe @bryanmacfarlane has an ETA? How much longer are we talking. 3 months? 6? 9? Some of us need to know in order to make decisions for our enterprise.

**Comment by axel-h** [comment]
Looks like this was supposed to get fixed at some point, but then it turned out this does not work as expected. I've read a bit into this `ACTIONS_RUNNER_PRINT_LOG_TO_STDOUT` feature now (because I'm still waiting for one of the test jobs to print another line, so I can see what they have done so far....), but I don't really get the point of having to set some flags to get a basic feature that should be there by default.
As a quick fix, why can't here be a link to click on, to get the intermediate log that as accumulated so far? It could have bandwidth restrictions, if load is really a concern here. Or just print the last 10/50/100 lines by default, that should be sufficient to give a hint if the job can be cancelled due to some error or if it is still within the expected progress.

**Comment by thaifp** [comment]
Cannot believe this issue was raised 2.5 years so far, and the github action team has not fixed it. I'm still painful when something wrong with my build, need to wait until it's timeout to find the cause.

**Comment by julienbonastre** [comment]
Pretty sad hey?

**Comment by Greg-Myers-SB** [comment]
Thread re-opened as https://github.com/actions/runner/issues/2622

**Comment by bryanmacfarlane** [comment]
It's implemented and sitting behind a feature flag. 😄   However, we're waiting on this: https://github.blog/changelog/2023-01-31-github-actions-job-summary-updates/    The work required the runner talking directly to a blob to get in progress buffers and since some customers have runners behind firewalls with firewall rules we had to update the documentation on what rules need to be set.  Since this alters that, a notice was required so their workflows don't break.  There is a huge focus on not breaking existing workflows - we would love to enable it ❤️ 

To answer the question on why we didn't show partial logs - there's two thing going on from the runner.   The runner writes lines to a web socket that users watching workflow from a web browser hook into.  So for in progress steps, you see lines from when you start watching.  At the same time, the runner is sending durable logs in largish MB pages to blob.  So even if we read existing pages, you wouldn't see what you're actually waiting on.

So the arch change is for the runner to write directly to blob with SAS tokens instead of funneling logs through our server endpoints.   So now when the web socket hooks up to read lines live, we can back fill from the blob pages being directly written to.

Hope that helps answering the questions above.

**Comment by charlesritchea** [comment]
Thank you for explaining and making progress, looking forward to the future

**Comment by adamscybot** [comment]
That is fantastic news! This is a huge usability improvement to GA. 

Would it be fair to say that July 31st is around about when this feature will land? Though if that feature flag could be exposed as an experimental flag to end users, that would be awesome.

**Comment by bryanmacfarlane** [comment]
That's certainly the goal right now and the current plan 😄  We're also heavily invested in reliability, availability and scale work.  We're scale testing the solutions and also making sure we don't introduce any reliability issues with this work and at the same time doing other arch work for scale/reliability.    We really want this as well and it's the main focus for one of our teams.

**Comment by julienbonastre** [comment]
Thanks @bryanmacfarlane , really appreciate the feedback on this one.. #github-actions-FTW ! 🚀

**Comment by allwalte** [comment]
Thanks for the update, @bryanmacfarlane !  However, I have a question re: https://github.blog/changelog/2023-01-31-github-actions-job-summary-updates/ that you linked - that's referencing what presumably are common/central URLs.  Will this feature be coming to GHE at some point, and would it still require that information go to central GH servers?

**Comment by ScottG489** [comment]
You also can't access logs from a job that **HAS** finished via the API if it's part of a workflow that hasn't finished even though **you can do this via the website**. Is this intentional?

This can be reproduced by running a workflow that has 2 or more jobs and after the first job has finished, run following `gh` command from within that workflows repo locally:

`
gh run view --log --job=$(gh run view $(gh run list --json databaseId --jq '.[0].databaseId') --json jobs --jq '.jobs.[] | select(.conclusion != "").databaseId' | head -1)
`

With the following output:
```
run <workflow id> is still in progress; logs will be available when it is complete
```
You can observe the problem more clearly by running:

`gh run view $(gh run list --json databaseId --jq '.[0].databaseId') --json jobs --jq '.jobs.[].conclusion'`
```
success

```
With the trailing empty line showing that the second job has yet to finish.

If you go to the workflow run in the browser, you can see the logs in the job and further from the job view you can even "view raw logs" that are served up from https://pipelines.actions.githubusercontent.com/serviceHosts.

These are further made inaccessible programmatically because it seems you need to be logged in to access the "view raw logs" link such as:
https://github.com/OWNER/REPO/commit/COMMIT_HASH/checks/JOB_RUN_ID/logs

**Comment by dillon-cullinan** [comment]
Just adding another use-case that is a problem here. We are currently using Github Actions Runner Controller in Kubernetes. This specific bug causes major issues when infrastructure hits resource limits. 

If a pod is terminated due to exceeding resource limits, you don't even receive the logs ever. You are stuck with a failure, logs don't load, and the only information that you have is the Summary page saying that the "self-hosted runner lost communication with the server".

The only way to debug such issues is to go into kubernetes logs, find the delete log, and figure out why the container exited. But this isn't perfect, as you don't see where it exited, you'll just see an exit code. You can only determine why it got deleted, but the only way to figure out where in the pipeline this happens is by having the build open at all times streaming your logs, which is a major flaw.

**Comment by ajschmidt8** [comment]
> It's implemented and sitting behind a feature flag. smile However, we're waiting on this: https://github.blog/changelog/2023-01-31-github-actions-job-summary-updates/ The work required the runner talking directly to a blob to get in progress buffers and since some customers have runners behind firewalls with firewall rules we had to update the documentation on what rules need to be set. Since this alters that, a notice was required so their workflows don't break. There is a huge focus on not breaking existing workflows - we would love to enable it ❤️

@bryanmacfarlane, any updates here now that we've reached the July 31st deadline in the post you mentioned above?

**Comment by ajschmidt8** [comment]
@dillon-cullinan, I've opened an issue for the problem you mentioned above here: https://github.com/actions/runner/issues/2387.

Unfortunately it was closed despite it still being a problem.

Also... hi :wave: :grin:

**Comment by Greg-Myers-SB** [comment]
@ajschmidt8 the issue is open here for tracking purposes.
 https://github.com/actions/runner/issues/2622
 I'll close it when I see it working as long as a collaborator doesn't do so first.

**Comment by rehanvdm** [comment]
Uhm yes please, why is this even an issue 😳. It's been 2 and a half years that GH has known and been sitting on this choosing not to fix it?

**Comment by bryanmacfarlane** [comment]
@rehanvdm see comment above: https://github.com/actions/runner/issues/886#issuecomment-1564345968

**Comment by rehanvdm** [comment]
Thank you, I saw but it does not change the fact that this basic functionality is missing from the GH core offering. I understand there are technical hurdles to overcome and that you can not just introduce changes easily but it has been more than 2 and a half years to find ways around it. But it seems that GH won't fix this? 

Won't it be better to just label this with "Won't fix" as well? Not to get everyone's hopes up?

Because the only other way around this is a DIY method?, which I really do not want to be responsible for. It would mean we have to intercept the standard streams from any long-running commands we run and send them somewhere else before handing them back to the standard streams to be interpreted and logged by GH.

**Comment by paulogodinhoaq** [comment]
@rehanvdm I am always thinking of going that route, but when we are doing that, why even use Actions in the first place...

**Comment by dnuzz** [comment]
Same, this is making some of our deployment processes effectively a black box

**Comment by mjperrone** [comment]
> The current ETA for beta testing this feature is end of August / early September timeframe.

@yacaovsnc, now that it is early September, are you able to provide an updated ETA?

**Comment by bryanmacfarlane** [comment]
if you want to try it out, then email me.  my email is my full name (alias) and that's at github dot com.  I can hook you up.   We finished scale testing and had a bug bash.  We're fixing a few bugs 🤞 so hopefully a few days.  I'm the messenger here and relaying back to the teams so 🙏 ❤️

**Comment by perryjrandall** [comment]
^ Amazing news! Us folks at aptos-labs would love to try it out if you can include us as well (sent an email too)

**Comment by Mause** [comment]
Any chance of getting this enabled on the duckdb and duckdblabs orgs? We have several long running jobs that would really benefit from this

**Comment by andrewakim** [comment]
:wave: Hi all - Actions PM here. I wanted to thank everyone for their interest in testing out the feature. We have a good amount of customers in the test group already so we won't be taking on anymore for now. I'll reach back out if we open it again.

**Comment by DARKo6o6o6** [comment]
If anyone wants to complete it help please

**Comment by will-lumley** [comment]
Very excited to test this feature.

**Comment by raphaccgil** [comment]
@andrewakim do we have any updates here?
We are using GA in our CI for DBT and in this case would important to have this information. 
In my scenario, we are using a DBT with Slim CI approach, this means we are using just a few data. However, the models still take a long time and we do not know if the CI is progressing or if any model has failed.

**Comment by andrewakim** [comment]
:wave: Hi all - Appreciate the patience. The team is still working on the feature and addressing a few issues we came across. I'll set a reminder for myself to post an update in 2-3 weeks and I'll hopefully have a bit more detail around timelines by then.

**Comment by andrewakim** [comment]
As promised, just wanted to share a quick update. We had to shift gears to focus on a few other priorities that came up in the last few weeks but we're planning on returning to this work later in the quarter. I'll keep sending out updates as things develop.

**Comment by humzam** [comment]
👋 Hi, @andrewakim @bryanmacfarlane 

What is the timeline looking like for this feature to be released?

**Comment by andrewakim** [comment]
:wave: Hi everyone - Good news! We're planning on rolling this feature out slowly starting in a few weeks. We'll begin to ramp up the rollout in the new year. Thanks for your patience!

**Comment by ayushxx7** [comment]
following...

**Comment by Dmitry1987** [comment]
asking for this feature is like asking Microsoft to give away % of their current revenue, literally.. the runner billing per minutes spent, duuh 😅 

Look: so there's a GET request "in the end when stage has been completed", which means they record the logs into some storage anyway, in the process, while the amazing websocket is "only streaming to browser". So letting the users access that "full log from current storage - whatever was cached so far" is a piece of cake (we're all developers here, and we *know it* 💯 ). The reasons why it would last 2.5 years to "fix" or implement the feature could be bandwidth cost (more GET requests in total, while stage is not yet completed) which is miniscule since we can stream the naughty websocket all day and watch logs anyway, even if we missed the critical first hundred lines - so bandwidth is not an issue. Or it could be the cost we all pay for the github runners per minute billed, and we obviously will spend more minutes waiting for stages to complete 🧐. Kinda ridiculous but since there are so many other pros to github, most of us keeps playing along like we don't understand what really goes on with this bug over all these years 😆 .  Enjoy that runners billing revenue, Microsoft! 🚀💰. Well we can't blame the engineering team, this is business 🤷‍♂️.

**Comment by brianjmurrell** [comment]
@Dmitry1987 Did you bother to read even just [2 comments](https://github.com/actions/runner/issues/886#issuecomment-1843569825) back before you posted your diatribe?

TL;DR: The fix for this is being rolled out.

**Comment by spaltrowitz** [comment]
There has been an update related to this public beta release: https://github.com/github/releases/issues/3686#issuecomment-1866738791

**Comment by brianjmurrell** [comment]
> There has been an update related to this public beta release: https://github.com/github/releases/issues/3686#issuecomment-1866738791

Bzzzzt.  That's a 404.

**Comment by spaltrowitz** [comment]
My apologies @brianjmurrell this post was in error. Please keep watching the roadmap for updates https://github.com/github/roadmap/issues/839!

**Comment by charlesritchea** [comment]
If you have "beta" badge on your run then log lines should update, if not
then it's still old behavior

**Comment by lure8225** [comment]
Thanks for the instant reply. I'm using a selfhosted runner via ARC. Will this issue also be resolved for these when this gets out of beta?

**Comment by charlesritchea** [comment]
I'm using a self hosted runner as well and have beta badge. I think it's
like a feature flag

**Comment by wrslatz** [comment]
Looks like this has been released! 🎉 
[Blog](https://github.blog/changelog/2024-04-30-github-actions-ui-improvements/), https://github.com/github/roadmap/issues/839
